File: features/snippets/external_storage/multiple_drivers/multiple_drivers.go
package multipledrivers

import (
	commonpb "go.temporal.io/api/common/v1"
	"go.temporal.io/sdk/converter"
)

// @@@SNIPSTART go-external-storage-multiple-drivers
type PreferredSelector struct {
	preferred converter.StorageDriver
}

func (s *PreferredSelector) SelectDriver(
	ctx converter.StorageDriverStoreContext,
	payload *commonpb.Payload,
) (converter.StorageDriver, error) {
	return s.preferred, nil
}

func MultipleDriversSetup(preferredDriver, legacyDriver converter.StorageDriver) converter.ExternalStorage {
	return converter.ExternalStorage{
		Drivers:        []converter.StorageDriver{preferredDriver, legacyDriver},
		DriverSelector: &PreferredSelector{preferred: preferredDriver},
	}
}
// @@@SNIPEND

File: features/snippets/external_storage/multiple_drivers/multiple_drivers.java
import io.temporal.payload.storage.ExternalStorage;
import io.temporal.payload.storage.StorageDriver;
import java.util.Arrays;

class MultipleDriversSnippet {
  static ExternalStorage configure(StorageDriver preferredDriver, StorageDriver legacyDriver) {
    // @@@SNIPSTART java-external-storage-multiple-drivers
    return ExternalStorage.newBuilder()
        .setDrivers(Arrays.asList(preferredDriver, legacyDriver))
        .setDriverSelector((context, payload) -> preferredDriver)
        .build();
    // @@@SNIPEND
  }
}
